Swarm command propagation, omq rpc, msg expiry & deletion

jagerman opened the pull request:
(Builds on top of #432)

This refactors RPC handling to unify it between HTTPS, onion requests, and (new) OMQ RPC requests, and to significantly clean up, abstract, and document what different commands do.

Breakdown of changes:
- Make random message retrieval more efficient (the current approach scales poorly as the database grows)
- Refactor how timestamps and TTLs are handled; we now store a timestamp and an expiry, but no longer the TTL value: instead the TTL value is simply a temporary value that can be given when storing a message to implicitly define the expiry.
- Remove dead code/disused OMQ endpoints.
- Fix SS not shutting down properly if it gets signalled while still trying to get the initial keys from oxend.
- Change database storage to store bytes rather than base64-encoded data.  Base64 is a transport encoding that is needed for javascript, but was being wastefully stored in the database too.
- Change generated hash to be based on fundamental values rather than user-provided encodings.
- Don't require timestamp and ttl to be passed as strings anymore.
- Allow storing using timestamp+expiry (as an alternative to timestamp+ttl).
- Expose storage rpc endpoints through a new public oxenmq rpc category `storage.WHATEVER` (e.g. `storage.store`).  This allows clients that want to use zmq speak more efficiently to SS.
- Add bt-encoded input/output support to the OMQ storage endpoints; this is noticeably more network efficient because it requires neither base64 encoding, nor establishing and handshaking new connections for every request.  The OMQ storage endpoints take params either json or a bt-encoded dict, and reply in kind.
- Clean up internals by moving transport encoding of internal values closer to the transport layer.  (Aside from a cleaner design, this was also needed to get bt-encoded responses out cleanly).
- Add delete/expiry API interfaces to delete all, deleted selected, delete by timestamp, shorten all expiries, and shorten specific message expiries.

Still TODO:
- [x] wire up actual deletion and expiry operations (currently just stubs)
- [x] propagate, collect, and coalesce responses from fellow swarm members
- [x] make the `store` call recursive (storage already is, but there is no confirmation and it should be adapted to use the new store mechanism + confirmations added herein)
- [x] add database migration the database (there are some unused columns, a completely broken index, some unfixable design problems relying on `rowid`, and some wasteful storage encoding).
- [x] add timestamp to all storage rpc returns so that the client doesn't have to fetch it unless they've been down for a long time.
- [x] more testing

jagerman committed:
Move all vendors cmake code to vendors/CMakeLists.txt

Also modernizes it a bit more to use cmake interfaces, plus fixes an
oxenmq linking issue.

jagerman committed:
Move server cert generation out of header

jagerman committed:
Clean up data_dir handling

string -> filesystem -> string -> filesystem is silly (and leads to
portability problems because strings can't be implicitly/successfully
converted to std::filesystem::paths on some systems).

Insert convert to a filesystem *once* (using u8path) and keep it there.

jagerman committed:
Replace boost::beast with uWebSockets

uWebSockets is very fast, has a much nicer interface than boost::beast,
and is currently used for handling all http rpc requests in oxen-core.

As part of this:

- Remove net stats singleton; it was very granular -- just number of
  http connections established with no indication of what type of
  connections (e.g. snode-to-snode or client) those are.  Will think
  about reintroducing some less granular stats gathering in the future
  if we need them.
- Update oxenmq to dev branch (for new cancellable timer support)
- move boost asio io_context into ServiceNode rather than main (and
  expose it publicly)
- add signal handler and shutdown logic
- remove long-deprecated (and non-functional) detection of long-poll
  headers
- import string_utils.h from oxen-core (for durations, int parsing,
  etc.); we already had *some* of this code with copies of
  starts_with/ends_with/join.
- import file.hpp from oxen-core (for file slurping)
- move more request logic into request_handler
- move common http types/values into `oxen::http` namespace (e.g. status
  codes, header names)

jagerman committed:
Rename lmq -> omq internally

jagerman committed:
oops

jagerman committed:
Fix --lmq-port backwards compat; test suite fixes

jagerman committed:
Banish windows linefeeds

jagerman committed:
Logger: use file/line instead of func

__func__ is pretty useless as all lambdas are reported as `operator()`
so you have no indication at all where the log msg is coming from.

jagerman committed:
Bump missing threshold to a weak inequality

When strict testnet *always* has to go out for help because the
threshold ends up at 0 (after integer division).

jagerman committed:
Improve logging output

- Show [filename:line] rather than __func__ because __func__ is useless
  when called from a lambda (it just shows `operator()`).
- Add time since startup (like lokinet does)
- Fix up log formatter so that it doesn't have to double-format and
  doesn't break when given a value that isn't a string literal for the
  format string.
- Calling via ->log instead of ->debug, etc. requires changing
  `OXEN_LOG(error, ...)` to `OXEN_LOG(err, ...)` to match the actual
  spdlog log level (which is different from the logging method for some
  reason).

jagerman committed:
Deal with boost::asio's event loop deficiencies

io_context run() returns immediately when there are no jobs added to the
ioc, and has no way to say "run until I call stop()" except by adding a
fake job into the io_context before calling run.

(Previously `run()` ran forever because we added the http server onto
it, which lasted forever).

jagerman committed:
https_server fixes

jagerman committed:
Move RequestHandler method implementations to request_handler.cpp

Having *some* of the method bodies for request_handler.h in
request_handler.cpp and others in onion_processing.cpp is rather
unexpected.

Also renames `is_server_url_allowed` to `is_onion_url_target_allowed`
(and rename the parameter from `url` to `target`) to better reflect what
it does and what it takes (i.e. it *doesn't* take a url).

jagerman committed:
Move parse_pubkey into oxend_key

Also add versions for each of the three pubkey types, and make it also
accept raw bytes (in addition to the existing base64/32z/16 values).

jagerman committed:
Remove boost beast & many intertwined changes

- Replace boost beast http(s) client code with cpr (same as used in
oxen-core and lokinet)

- Replace bootstrap http requests with encrypted/authenticated omq rpc
requests to the bootstrap nodes (rather than unauthenticated,
unencrypted http connections).

- Remove boost::asio

- Remove boost circular buffer; a regular map with a two-line trimming
code is simpler for block hash cache, and a limit on stored snodes
doesn't seem necessary for the rate limiter.

- Make rate_limiter clean itself periodically; currently it only cleans
once it reaches max clients (i.e. 10k).

- Make rate_limiter thread safe, plus various code cleanups/DRYing.

- Remove ip_utils; we now longer really need a check for public ip for
proxy requests because we are sufficiently restrictive on the proxy url
target that we are unlikely to be able to reach anything sensitive (and
we also don't allow redirects).

- Replace /swarms/ping_test_v1 with /ping_test/v1; this new request now
returns the remote pubkey in a header, and no longer includes an SSL
cert signature (so that we can drop the SSL cert signatures after HF19).
The old one will still be used until HF19.

- Add OMQ endpoint for storage tests; it gets used instead of the https
one at HF19.

- Refactor storage test retries into request_handler (needed so both
legacy and omq endpoints can use it).

- Fix HF18 swarms/storage_test endpoint which was missing the snode
signature required for current SS versions to accept the response.

- Move HTTPS server-specific code (validate_snode_signature) out of
RequestHandler into HTTPSServer

- Make onion proxy-to-url timeout a bit less than the onion request
timeout so that the client has a better chance of actually getting a
timeout error.

- Miscellaneous cleanups

- Remove pre-HF18 compatibility code.

- Remove `process_lns_request` onion request target: it was accidentally
broken in HF18; clients are now using an oxend rpc request proxy to
`ons_resolve` instead.

- Shorten timeout values for ping and storage tests (5s and 15s,
respectively) and bootstrap connections (now 10s instead of 60s).

jagerman committed:
Convert unit tests to Catch2

It is much nicer than boost or gtest, especially when running tests, and
is what we use very successfully in oxenmq and lokinet for unit tests.

jagerman committed:
Add check for up-to-date submodules

jagerman committed:
Add HTTPS ping test debug

jagerman committed:
Fix reversed logic on https ping test url

jagerman committed:
More debug

jagerman committed:
Fix bad url

jagerman committed:
Slow down testing tick timer

20 times per second seems much faster than is useful.

jagerman committed:
Stats cleanup and refactor

Refactors how stats are collected:

- Keeping a rolling window of recent stats for every 10 minutes over the
past hour so that when we rotate stats, we only chop off the 1h10min-1h
counts, but always have a recent stats counter that reflects the last
~1h of requests.

- Move the stats cleanup timer into the stats class itself

- Make the all_stats class properly thread safe.  Previously *parts* of
it were thread safe but much of it wasn't, so really wasn't safe to be
used without holding the owning service_node's global lock.

- Fix various bugs in the stats values:
    - "reset_time" was never updated, but even if it had been, it was a
    steady_clock and being returned as time_since_epoch() which is a
    useless value.

    - recent_retrieve_requests wasn't being set at all in the returned
    json (instead recent_store_requests was set twice).

    - onion requests weren't being counted at all; instead every onion
    request bumped the proxy request counter

- Got rid of exposing "current period" stats: instead we now just expose
the recent request counters, which are calculated from current period +
stored historic period values, thus giving you a count for requests over
the last 1h5m (±5 minutes).

- Replace time_t with modern time types

- Don't indent/prettify json; in any case where you'd actually want it
this is much better done via tools like `jq`.

jagerman committed:
Better status line (for systemd Status)

- Show swarm value in hex instead of decimal
- Add onion/proxy values
- Use recent stats instead of global stats

jagerman committed:
Fix onion/proxy request recording

Proxy requests weren't being recorded at all; onion requests were being
recorded in an awkward location.

jagerman committed:
Fix use-after-move crashes

jagerman committed:
Fix how cmake finds spdlog

jagerman committed:
Build static jemalloc

jagerman committed:
Fix recent stats window being 10min too large

jagerman committed:
Add used page count stats

jagerman committed:
Database: use logger

Raw printing to stderr is gross and racey.

jagerman committed:
Database: enable WAL

Because it's better in just about every way.

jagerman committed:
Fix stats showing wrong store reqs value

jagerman committed:
Remove buffered message relaying

It's counterproductive (messages are already going out via omq) and so
just ends up wasting memory and making swarm propagation slower.

jagerman committed:
Increase swarm batch serialization size to 9MB

Batch serialization is now used only when a SN first joins a swarm; in
that case there's really no advantage of many small bundles so just bump
up the size closer to our max message size.

jagerman committed:
Build fixes

- add autoconf dep for jemalloc
- add libuv for macos

jagerman committed:
Remove unused Findsodium cmake script

jagerman committed:
jemalloc build fix

jagerman committed:
Drone: remove boost deps, add curl deps

jagerman committed:
Need make too

jagerman committed:
Fix batch serialization unit test for new larger batches

jagerman committed:
jemalloc build fixes for older cmake compat

jagerman committed:
Status line tweaks

- change "a/b/c/d in 61.7min" -> "a/b/c/d (last 61.7min)"
- reduce db size precision by 1 (e.g. 41.7MB instead of 41.68MB)
- account for rounding so that we show 1.00GB rather than 1000MB for
values in [999.5MB, 1GB) that will get rounded up (and similar for
1000kB)

jagerman committed:
Allow using system jemalloc for non-static build

jagerman committed:
Add curl to static deps build

jagerman committed:
Add missing unordered_map header

jagerman committed:
Optimize random message retrieval

The approach being used here with an offset is painfully inefficient,
and has a race condition; this switches it to something better.

This also allows elimination of the ridiculous
"util::uniform_distribution_portable" call which didn't produce anything
portable at all.

jagerman committed:
Replace user-agent on outgoing http requests

If unspecified then the libcurl version gets added.

jagerman committed:
Refactor how TTL & timestamps are stored

Currently we have an awkward storage of timestamp + ttl + expiry in the
database, and timestamp + ttl passed in from the client (as strings!).
Althis is awkward because we want to be able to shorten the expiry, but
that would mess up TTL.  Additionally everything is stored as
`uint64_t`s, which is messy and not type safe.

This commit makes these changes:

- ttl and timestamp can now be sent by the client as integers (in
addition to the current string value).  NB: this is not reliable until
the entire SN network is on the next SS release.

- ttl, timestamp, and expiry are now type-safe std::chrono types instead
of raw integers (milliseconds for ttl and system_clock::time_points for
the other two).

- ttl is no longer stored: instead we just store timestamp + expiry.
(This will let us update expiry later without worrying about TTL).

- ttl/timestamp value validation is moved out of `common/oxen_common.h`
(which was a very odd place for it) and into request_handler.

- serialization no longer supports message_t; rather message_t is *only*
for holding the value the client gives us.  SS now uses storage::Item
everywhere other than incoming client data.

- Removed TTL/Nonce storage and retrieval

- Fully specify the queries columns instead of using `SELECT *`

- Don't use "`" for identifier quoting inside sqlite.  It's non-standard
MySQL garbage that sqlite3 supported only for MySQL compatibility.

jagerman committed:
Replace unsafe static local vars with atomics

jagerman committed:
Add FIXME comments around some broken database queries

jagerman committed:
Remove -Wfatal-errors

It makes editing the project incredibly annoying because you can only
get one error at a time (e.g. in ycmd) instead of many errors at once.

The comment here nor the commit message that added it don't justify
adding it in any way, so just remove it.

jagerman committed:
Refactor user_pubkey_t

Not having it default constructible is a pain, and the static `create()`
method is poorly named.

Refactor it to be default constructible, to have a `load` method that
loads from a string, and to have a `bool` operator that returns true
only after a valid pubkey has been loaded.

jagerman committed:
Remove dead omq endpoints sn.proxy_exit and sn.onion_req_v2

`sn.proxy_exit` is dead code (it is only callable from other SNs, and
isn't actually being called at all).

`handle_onion_req_v2` was for the old (pre-HF18) onion requests
(`sn.onion_req_v2`), which isn't used since the HF18 hardfork (and the
invoking endpoint registration has already been removed).

jagerman committed:
RPC: refactor and wire up to OMQ

Refactors storage rpc requests to abstract parsing and then wires them
up so that all of direct, onion-requested, and omq can now take the same
codepaths for requests.

The new OMQ requests, in particular, are publicly accessable at
`storage.whatever` endpoints (e.g. storage.store).

This also relaxes/changes some of the argument parsing:
- allow `pubkey` to be passed as that (currently `pubKey` is required,
and continues to work)
`last_hash` (AKA `lastHash`) is no longer required; omitting it is now
the same as providing an empty one.

Data is now stored as binary in the database; previously we were storing
the base64-encoded value received from the client.  (This will, however,
break any client that expected to be able to send random data).

Also adds an `info` storage endpoint that returns the current version &
timestamp.

TODO: need a database migration to convert existing (base64) data.  (We
need a db migration for other reasons, as well).

jagerman committed:
Make SS respond to signals when trying to get the initial keys

Without this it just loops forever and ignores the signal.

jagerman committed:
Update hash test and hashing construction

This changes how we compute hashes to better match how data is now
stored (i.e. expiry instead of ttl, binary data instead of base64).

This results in different hashes across upgraded/non-upgraded storage
servers, however Session currently handles that properly (and will
detect the duplicate) so while this will temporarily result in some
double storage during the migration period, it shouldn't cause any
other issues.

jagerman committed:
Add bt-encoded OMQ RPC interface

This makes client requests directly available over OMQ (using
bt-encoding rather than json) and will also be used for forwarded
requests between SNs (again using the more efficient bt encoding).

jagerman committed:
OMQ RPC: allow omitted params

For commands (like storage.info) that don't require any parameters.

jagerman committed:
Tweak TTL validation and error handling

jagerman committed:
Support bt-encoded response (needed for forwarding)

When an incoming OMQ RPC storage.whatever client request is bt-encoded
we respond in kind.  This isn't all that critical for clients, but more
important this is required for bt-encoded comms between SNs, which is
very useful, particularly when synchronizing messages (bt-encodes bytes,
so is 75% of the size of base64 encoded values).

This required moving the json "dump()" call much closer to the response;
this lets us convert the json into bt_value in the protocol handler,
rather than pushing dumped json through the interface.

Aside from enabling bt-encoded response, this is also cleaner design.

jagerman committed:
Add delete/expiry and swarm propagation stubs

majestrate commented on crypto/src/oxend_key.cpp:
```diff
@@ -57,4 +60,34 @@ x25519_pubkey x25519_seckey::pubkey() const {
 };
 
+template <typename T>
+static T parse_pubkey(std::string_view pubkey_in) {
```
return `std::optional<T>` ?

jagerman replied:
Not necessary for the pubkey types: they have a built-in `operator bool()` that returns false if set to an all-0 value.

majestrate commented on httpserver/service_node.cpp:
```diff
@@ -833,108 +851,121 @@ void ServiceNode::oxend_ping() {
+                    if (r.error.code != cpr::ErrorCode::OK)
+                        OXEN_LOG(debug, "FAILED storage test of {}: {}", pk, r.error.message);
+                    else if (r.status_code != 200)
+                        OXEN_LOG(debug, "FAILED storage test of {}: received non-200 status {}",
+                                pk, r.status_code, r.status_line);
```
missing `{}` or extra formatting argument

jagerman committed:
Add missing status line argument

jagerman committed:
DRY out database & switch to smart pointers

Removes a bunch of repetition; switches raw pointers to smart pointers
for database and prepared statements.

jagerman committed:
Update required sqlite3 to 3.35.0

This lets us use RETURNING for updates/deletes which makes things much
nicer.

jagerman committed:
Database: implement delete/expiry-update calls

jagerman committed:
WIP: Wire up delete/expire rpc endpoints, part 1

Exposes the rpc functions and adds a preliminary distribution between
SNs.

- Adds bt-to-json conversion and expose the json-to-bt converter for us
with converting incoming json requests to bt-encoding (for inter-swarm
relaying), and back (for responding).
- Wire up database functions <-> service node calls <-> rpc endpoints.
- Add swarm command distribution.

(This is not yet working).

jagerman committed:
Remove idea of passing the client IP

It doesn't make sense, particular when we are typically coming via onion
requests and won't know the client IP anyway.

jagerman committed:
Don't use sodium piecewise ed25519 construction

It doesn't work as expected: rather it creates an "ed25519ph"
signature, which is a signature over a pre-hashed value, rather than a
regular Ed25519 signature.

Instead this now uses an allocation-minimized conversion/concatenation
of arguments to build the proper Ed25519 signature message.

jagerman committed:
Fix pending, add request debugging

jagerman committed:
Store relayer's own result in "swarm" dict when recursive

Otherwise it's a bit weird to have one result at the top-level and the
others under "swarm".

jagerman committed:
Remove unused constant

jagerman committed:
Make get_results return # rows, -1 on error

This lets it be used for queries that might (legitimately) return 0
rows.

jagerman committed:
Fix forward request handling

jagerman committed:
Document response codes

jagerman committed:
Fix "Store relayer's own result"

jagerman committed:
Change message hash to blake2b-256 + base64

SHA512+hex is *way* longer than needed for a message identifier and is
just wasteful.

jagerman committed:
Reduce swarm propagation timeout to 5s

jagerman committed:
Check request being for us before replying

jagerman committed:
Wire up remaining delete/expiry endpoints

- Further documents each endpoint

- Changes the expiry update endpoints to just return the hashes of
messages rather than hashes + timestamps as its quite a bit simpler
(especially for signing).

- implements delete_msgs/delete_before/expire_all/expire_msgs endpoint
processing logic.

jagerman committed:
Add method name to signature generation

Without this you could submit an expiry request and someone could replay
it later as a delete request (when the expiry timestamp arrives).

With the method include this isn't possible.

jagerman committed:
Fix message hash acceptance

jagerman committed:
Include timestamp in retrieve request

jagerman committed:
Fix: valid message hash

majestrate commented on httpserver/client_rpc_endpoints.h:
```diff
@@ -0,0 +1,331 @@
+/// or stringified integer, relative to the timestamp.  Timestamp+ttl must not be in the past.  For
+/// backwards compatibility may be passed as a stringified integer.
+/// - `expiry` (required, unless ttl given) the message's expiry time as a unix epoch milliseconds
+/// timestamp.  (Unlike the above, this cannot be passed as an integer).
```
as a stringified integer

jagerman committed:
Add SQLiteCpp

This is a nice C++ wrapper that cleans up the data interface
considerably over using the C sqlite3 API.

I also evaluated (and started implementing) sqlite_orm for this, but ran
into considerable obstacles: the orm components get in the way without
being good enough to really solve anything (and essentially just making
you write queries in C++ code that is much less elegant than straight
queries), plus it fundamentally doesn't support threaded operation,
which sucks.

jagerman committed:
Fix/update boost mirror URL

boost's mirror moved.

jagerman committed:
SS refactor/cleanups

- Eliminated storage::Item because it just duplicates message_t for no
useful reason.

- Always pass user pubkeys as user_pubkey_t, never as std::strings.
Pubkeys are not strings, and the same pubkey has multiple string
representations.

- Properly split up the pubkey as [type][pubkey] where type = byte (5
for Session) and pubkey is a type-dependent key (currently only
supporting 32-byte Ed25519 keys).

- Allow pubkeys to be loaded from either hex (66) or binary (33) values.

- Allow pubkey prefixes on testnet.  A 32-byte key will be interpreted
as if the key was prefixed with a 0.  Thus there is now just one
"proper" pubkey size of 33 bytes or 66 bytes, and we silently accept the
missing-prefix for testnet (but can do away with the proper sizes being
different).

- Expose pubkey stringification in different ways (prefixed/unprefixed,
hex/raw).

- Simplify time_point-to-integer conversion with new time.hpp functions
from_epoch_ms and to_epoch_ms.

- Restore old message hash format so that we can keep using it until the
mandatory transition point.

- Added a much more network efficient serialization format, using
standard b-encoding rather than NIH custom encoding.  After the
mandatory upgrade we can remove the old format.

- Refactor swarm bootstrap code to be more efficient

jagerman committed:
Database redesign, refactor & migration

Redesigns the database to be a more appropriate, less duplicative design
using "owners" and "messages" with a foreign key between them.

Rewrites all the database code using SQLiteCpp which substantially
reduces the amount of boilerplate, duplicate code for query handling.

Makes the statement handlers thread_local for better thread safety; this
also allows the actual query to be where it is executed, rather than
having all the prepared queries in one place nowhere close to where they
are actually used.

jagerman committed:
Make `store` recursive

This allows a client to store on *any* swarm member, have the swarm
member distribute, and then verify the result of the store (if it wishes
to).

jagerman committed:
Make unit test logger go via Catch2 UNSCOPED_INFO

Hides log spam while still allowing Catch2 to show logs leading up to a
failure when a test failure occurs.

jagerman committed:
Test suite fixes for recent changes

jagerman committed:
Bump version to 2.2.0

jagerman committed:
C++ improvements

Remove superfluous `this->`.  This is C++, not Python/Rust/etc. that
require `self` boilerplate.

Use optional/pair for optional multi-value return rather than mutable
lvalue arguments with a bool return (which wasn't even checked in one
place).

jagerman committed:
Remove `_t` suffix from concrete structs

The `_t` suffix stands for `typedef` and is deceptive when used for a
concrete type.

jagerman committed:
Remove unused/uncommitted header

jagerman committed:
Rename badly named hex_to_u64, document it, and test it.

jagerman committed:
Add swarm (as hex string) into swarm members response

jagerman committed:
Print swarmid info message in hex

jagerman committed:
SQLiteCpp: Disable column metadata

We don't use it, and it isn't always available in sqlite3.

jagerman committed:
DB migration fixes

jagerman committed:
Fix variant error on string proxy callback

jagerman committed:
Use 500 status if all answers are failures

jagerman committed:
Fix store response not being set in swarms

jagerman committed:
If we get a recursive failure put the message in "reason"

jagerman committed:
HF version gating & store distribution fix

- Improve serialization versioning constants

- Use constants for version gating when `store` becomes recursive

- Don't do a background message distribution when we are using recursive
stores.

jagerman committed:
Version-guard blake2b hash

jagerman committed:
Add support for minor forks (spoons? sporks? snode revisions?)

jagerman committed:
Re-add seed1/seed3 to seed node list

jagerman committed:
Remove testnet exceptions to HF guards

jagerman committed:
Fix duplicate failing enqueuing

Nodes were being removed from `failing` while queued for testing, which
could result in multiple tests of the same node getting queued.

jagerman committed:
typedef the clock, not the clock's time

jagerman committed:
Update oxend endpoint for reachability reporting

jagerman committed:
get_swarm: +port_https/port_omq; deprecate "port"

Not only is it only providing one of the ports, it also stringifies it
for no good reason.

jagerman committed:
Add basic swarm testnet test suite

jagerman committed:
Fix pubkey encoding in https header

jagerman committed:
Fix last_hash parsing to accept blake2b+b64 hashes

jagerman committed:
More test suite

jagerman committed:
Work around Session not exposing ed25519 pubkeys

Session ids are, unfortunately, x25519 pubkeys rather than ed25519, and
there isn't a one-to-one conversion from x to ed, so we have to have the
client send along the ed25519 pubkey *along with* the 05-prefixed x25519
pubkey, then we also check that the given ed25519 converts to the x25519
pubkey when checking client signatures.

jagerman committed:
Make sure SS https is listening exclusively on the port

jagerman committed:
Explicitly mention that returned hashes are sorted

jagerman committed:
Bug fixes

- user pubkey hex prefix addition had inverted netid logic on testnet
- message testing was failing an assertion in debug mode

jagerman committed:
Allow ed25519 pubkeys for unprefixed addresses on testnet

jagerman committed:
Add delete endpoint tests

jagerman committed:
Add expire tests

Fixes wrong signature generation in server-side `expire` endpoint.

jagerman committed:
Fix storage tests (v2)

The previous fix wasn't enough; we need to properly handle both 64 and
32-bit keys both when sending the test request, and when handling that
test request.

jagerman committed:
Fix propagation of session ed25519 pubkey

The pubkey_ed25519 wasn't getting propagated when forwarding to swarm
members.

jagerman committed:
Add session ed25519 pubkey hack test

Tests support for Session's rather awkward use of the x25519 pubkey
in session ID rather than ed25519 pubkey.

jagerman committed:
Use constants for tolerance checks

jagerman committed:
Add delete_all/expire_all signature timestamp tests

jagerman committed:
network-tests: Add readme; rename to `network-tests`

jagerman committed:
Change unprefixed netid to 05 for testnet instead of 00

Better Session testnet compatibility.

jagerman committed:
Make startup more synchronous; oxenmq update

This does better synchronization at startup:

- OMQ listening ports now aren't opened until we have processed a block
update from oxend.  (Fixes #358)
- HTTPS is similar delayed
- port status messages are now shown when opening, instead of from
main().

This updates oxen-mq to latest dev (needed to support after-start()
listening), which also brings a potential connection mismatch fix.

jagerman committed:
Disable ephemeral routing id for SN-to-SN connections

This was supposed to be off all along, but I think got switched to true
when the badly named oxenmq variable changed.

jagerman committed:
g++-11 compilation fix

jagerman committed:
Remove superfluous trailing return types and obfuscating typedef

- trailing return types in C++ are for complex, dependent return types;
they are more verbose and less useful for plain return types.

- don't use a typedef for a relative simple type because it obfuscates
the simple type.

jagerman committed:
Allow payload/control input from files

jagerman committed:
Set content-type for post requests

Without this requests were getting sent with
'application/x-www-form-urlencoded' which completely broke the PN
server.

jagerman committed:
Fix segfault in stats cleanup

jagerman committed:
Fix submodule repo url

jagerman committed:
Remove embedded jemalloc submodule

The system jemalloc available on all the systems we support seems to
work well enough.

jagerman committed:
Add user count to status line

jagerman committed:
Add hourly stats to log file

jagerman committed:
Avoid forwarding "store" requests before HF18.1

Other nodes fail with UNKNOWNCOMMAND:sn.storage_cc

jagerman committed:
Remove removed submodule from submodule check

jagerman committed:
Update oxen-mq to fix connection error

jagerman committed:
Preload recent block hashes at startup

This is needed to avoid "We have no snodes to derive tester/testee from"
errors for the first few blocks after startup.

jagerman committed:
Fix logic on block polling

jagerman committed:
Fix legacy http storage tests: base64 encode the data

Older storage servers store everything as base64, so when they send the
response, they are sending out base64.  We were sending it as-is, but
that often makes json throw because our raw data isn't valid utf-8.

Encode/decode base64 to fix interoperability, and fix the json issue.

jagerman committed:
Update oxenmq

jagerman committed:
Fix infinitely recursive store requests

jagerman committed:
Add optional retrieve authentication

This is transitional: we don't require it yet, but if you provide the
auth fields we check it (and fail when invalid).

In the future, we will make authentication required, so in the meantime
clients that upgrade to use it won't break when the flip gets switched
later on.

jagerman committed:
Add authenticated retrieve network test

jagerman committed:
Remove unused fwd decls; fix class/struct mismatch

jagerman committed:
Bump oxenmq to stable 1.2.6 release

jagerman committed:
Update spdlog to latest stable version

jagerman committed:
Avoid alias to imported global lib for older cmake compatibility

jagerman committed:
Disable LTO for sqlite3+clang

jagerman committed:
Properly set include dirs for locally build sqlite3

jagerman committed:
Add timestamp to all client rpc responses

Might make clients life a bit easier by not having to go through the
"info" endpoint when they have some other recent request.
